src/recon_buildblock/LogcoshPrior.cxx

```diff
@@ -0,0 +1,598 @@
+          weights[0][0][0] = 0;
+        else
+        {
+          weights[z][y][x] =
+                  grid_spacing.x()/
+                  sqrt(square(x*grid_spacing.x())+
+                       square(y*grid_spacing.y())+
+                       square(z*grid_spacing.z()));
+        }
+      }
+}
+
+template <typename elemT>
+double
+LogcoshPrior<elemT>::
+compute_value(const DiscretisedDensity<3,elemT> &current_image_estimate)
+{
+  if (this->penalisation_factor==0)
+  {
+    return 0.;
+  }
+
+  const DiscretisedDensityOnCartesianGrid<3,elemT>& current_image_cast =
+          dynamic_cast< const DiscretisedDensityOnCartesianGrid<3,elemT> &>(current_image_estimate);
+
+  if (this->weights.get_length() ==0)
+  {
+    compute_weights(this->weights, current_image_cast.get_grid_spacing(), this->only_2D);
+  }
+
+  const bool do_kappa = !is_null_ptr(kappa_ptr);
+
+  if (do_kappa && !kappa_ptr->has_same_characteristics(current_image_estimate))
+    error("LogcoshPrior: kappa image has not the same index range as the reconstructed image\n");
+
+  double result = 0.;
+  const int min_z = current_image_estimate.get_min_index();
+  const int max_z = current_image_estimate.get_max_index();
+  for (int z=min_z; z<=max_z; z++)
+  {
+    const int min_dz = max(weights.get_min_index(), min_z-z);
+    const int max_dz = min(weights.get_max_index(), max_z-z);
+
+    const int min_y = current_image_estimate[z].get_min_index();
+    const int max_y = current_image_estimate[z].get_max_index();
+
+    for (int y=min_y;y<= max_y;y++)
+    {
+      const int min_dy = max(weights[0].get_min_index(), min_y-y);
+      const int max_dy = min(weights[0].get_max_index(), max_y-y);
+
+      const int min_x = current_image_estimate[z][y].get_min_index();
+      const int max_x = current_image_estimate[z][y].get_max_index();
+
+      for (int x=min_x;x<= max_x;x++)
+      {
+        const int min_dx = max(weights[0][0].get_min_index(), min_x-x);
+        const int max_dx = min(weights[0][0].get_max_index(), max_x-x);
+
+        /* formula:
+         sum_dx,dy,dz
+         weights[dz][dy][dx] *
+         log(cosh(current_image_estimate[z][y][x] - current_image_estimate[z+dz][y+dy][x+dx])) *
+         (*kappa_ptr)[z][y][x] * (*kappa_ptr)[z+dz][y+dy][x+dx];
+         */
+        for (int dz=min_dz;dz<=max_dz;++dz)
+          for (int dy=min_dy;dy<=max_dy;++dy)
+            for (int dx=min_dx;dx<=max_dx;++dx)
+            {
+              // 1/scalar^2 * log(cosh(x * scalar))
+              elemT temp = current_image_estimate[z][y][x] - current_image_estimate[z+dz][y+dy][x+dx];
+              elemT current = weights[dz][dy][dx] *
+                      (1/(this->scalar*this->scalar))*log(cosh(this->scalar*temp));
+
+              if (do_kappa)
+                current *= (*kappa_ptr)[z][y][x] * (*kappa_ptr)[z+dz][y+dy][x+dx];
+              result += static_cast<double>(current);
+            }
+      }
+    }
+  }
+  return result * this->penalisation_factor/2.0;
```

Comment:
I was a little surprised to find that this is needed. The test of #695 only passes with this halving.

Reply:
this is the problem that Tsai found. It appears because it's a derivative of a sum of symmetric functions. Similar factors can appear in the Hessian and surrogate.

Reply:
We have no way of testing for those right now. I think with the current implementations, the hessian values need to be multiplied by 2. 

The surrogate is not as obvious for me.